### ISSUE TITLE
Linting Issues & Questions

### ISSUE THREAD

**Issue by zrosenbauer** [neighborhoods/Facade-Inspector#2]
`.eslintrc` from Boilerplate...

``` json
{
    "import/prefer-default-export": 0,
    "padded-blocks": 0,
    "react/jsx-filename-extension": 0,
    "import/no-extraneous-dependencies": 0,
    "import/extensions": [0, {
      "js": "always",
      "jsx": "always",
      "json": "never"
    }],
    "no-underscore-dangle": 0,
    "space-before-function-paren": 0
 }
```
- import/prefer-default-export => prefers a default export per file, which breaks with our action creator files
- padding-blocks => I think should be up to the dev
- react/jsx-filename-extension => only issue is the client.js file that contains some XML but is not truly a react file
- import/no-extraneous-dependencies => redux-mock-store etc. not sure on issue here but seems a nuisance, will look into more this week
- no-underscore-dangle => breaks any of our system global `__SERVER__ || __DEVELOPMENT__`
- space-before-function-paren => do we want to support this? I prefer `function()` `vs function ()` I think we should change the setting to "never" which would allow only the `function()`

**Comment by jrubins** [comment]
Addressing @zrosenbauer's points:
- import/prefer-default-export
  - This should not break with our action creator files - it should only apply if you have a single action creator (or a single export in a module).
- padded-blocks
  - I personally don't like these because it leads to messy and inconsistent spacing, but I don't feel strongly about it.
- react/jsx-filename-extension
  - I'm up for getting rid of this rule. I've never actually named my files JSX since I think the structure of the app (components directory) mostly indicates where JSX files are.
- import/no-extraneous-dependencies
  - I like this rule. We shouldn't be trying to import something that isn't in our `package.json` or that's not a relative path to one of our own modules.
- no-underscore-dangle
  - We could add these globals as exceptions. I think I'd prefer that but if you guys feel strongly we can get rid of this rule. It also is an annoying rule when you're working with variables outside of your control (like things that come back from other APIs, although I'd prefer to just disable the rule for that line).
- space-before-function-paren 
  - I also prefer `function()` over `function ()`. I agree with updating to "never".

Additional thoughts from fixing NHDS Facade issues:
- no-console 
  - Should we get rid of this? I actually think it's a good thing to have since using `console.*` server side is synchronous. But it would take a little bit of refactoring.
- no-return-assign 
  - Should we keep this? That means our ref callbacks will have to look like: `ref={(ref) => (this.ref = ref)}` instead of `ref={(ref) => this.ref = ref}`.
- no-unused-expressions 
  - I think we should disable this since it doesn't seem to pick up on our short circuit expressions well (which we use all over the place in NHDS Facade)

cc @cbourdage @VBurroughs21 @clermontster @MattStypa

**Comment by jrubins** [comment]
I also vote we turn on the `MemberExpression` option for the `indent` rule. See: http://eslint.org/docs/rules/indent

**Comment by zrosenbauer** [comment]
@jrubins 

# I'd like to keep..
- no-return-assign
  - I like this in that it will keep us from accidental assignments vs conditionals (which should be caught in PRs) but we could drop to a warning OR just use the `(ref)=>{ this.ref = ref }` syntax, I'm not opposed to that..
- no-unused-expressions
  - Seems to catch unused modules etc. pretty well I'm all for keeping this as it could cut down on our build size if we're including modules that aren't needed (legacy, accidental etc.)
- import/no-extraneous-dependencies 
  - only caused issue in tests but adding `jest` to eslint should resolve
- no-underscore-dangle 
  - I personally dislike the underscore dangles so I'm def on board with keeping this, if we can 

# I'd like to remove..
- padded-blocks
  - +1 to remove this then?
- react/jsx-filename-extension
  - I'm actually fine with that also, the compiler will resolve js vs jsx without extensions needed
    add in those exceptions per project then I'm fine with it

# Needs an update
- space-before-function-paren
  - update to never! 😄 

# I have questions...
- import/prefer-default-export 
  - That's fine but since we are preferring sub-module exports vs default exports would we recommend disabling that rule / eslint in certain modules, say in a util file that has only one function but more will be added long term.
- no-console
  - I'm fine with keeps this, I'm not a fan of littering the console with random information but we also used a log wrapper such as `Logger::error` etc. where we could disable the linting but still keep it in the project, also allows us to post errors to an endpoint in the future, handle browsers that don't support full console object etc.
- consistent-return
  - I'd like to add this to the default settings [rule](http://eslint.org/docs/rules/consistent-return)
  
  ``` javascript
    "consistent-return": [2, {
      "treatUndefinedAsUnspecified": false
    }]
  ```

**Comment by jrubins** [comment]
- no-return-assign
  - I think we should just use the "except-parens" option which will make our format:

```
(ref) => ( this.ref = ref )
```
- no-unused-expressions
  - You may be confusing this with `no-unsued-vars` - `no-unused-vars` will error if you don't use an imported module. `no-unused-expressions` isn't great at picking up on the following (which is why I think we should remove):

```
myCondition && this.props.doAFuncCall()
```
- padded-blocks
  - I think we should keep this enabled - I'm not a big fan of coming into a file and seeing blocks padded in a bunch of different ways. The following scenarios bug me:

```
if (thisIsTrue) {

  doThis();

} else if {
  thisOtherThing();

} else {

  thisThing();
}
```

vs.

```
if (thisIsTrue) {
  doThis();
} else if {
  thisOtherThing();
} else {
  thisThing();
}
```
- import/prefer-default-export
  - That's a good point. I think I'd prefer to turn this rule off then. I'm not a fan of disabling linting for files since it will lead to forgetting to turn it back on.
- consistent-return
  - That option is the default... are you having an issue?

**Comment by jrubins** [comment]
Others to consider removing:
- react/no-did-mount-set-state
- react/no-did-update-set-state

**Comment by zrosenbauer** [comment]
@jrubins we want or need to support set state with-in the ComponentDidMount and ComponentDidUdate methods?

**Comment by jrubins** [comment]
@zrosenbauer Yes, we want to allow setting state in `componentDidMount` or `componentDidUpdate` is my proposal. Right now, you're not allowed to.

**Comment by zrosenbauer** [comment]
I'm fine with that add it in.

**Comment by zrosenbauer** [comment]
@jrubins did we ever get a PR for this open and merged in?

**Comment by jrubins** [comment]
Closed by #5
